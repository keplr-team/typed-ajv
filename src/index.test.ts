import { CS } from '.';

const checkType = <A>(a: A) => a;

it('works with string schema', () => {
    const str = CS.String(true);
    type strType = typeof str.type;

    checkType<strType>('abcd');
    expect(str.getJsonSchema()).toEqual({
        type: 'string',
        transform: ['trim'],
    });
});

it('works with number schema', () => {
    const cs = CS.Number(true);
    type type_ = typeof cs.type;

    checkType<type_>(2);
    expect(cs.getJsonSchema()).toEqual({
        type: 'number',
    });
});

it('works with integer schema', () => {
    const cs = CS.Integer(true);
    type type_ = typeof cs.type;

    checkType<type_>(2);
    expect(cs.getJsonSchema()).toEqual({
        type: 'integer',
    });
});

it('works with boolean schema', () => {
    const cs = CS.Boolean(true);
    type type_ = typeof cs.type;

    checkType<type_>(true);
    expect(cs.getJsonSchema()).toEqual({
        type: 'boolean',
    });
});

it('works with any schema', () => {
    const cs = CS.Any(true);
    type type_ = typeof cs.type;

    checkType<type_>(2);
    checkType<type_>('abc');
    checkType<type_>({ a: 1 });
    expect(cs.getJsonSchema()).toEqual({});
});

it('works with object schema', () => {
    const obj = CS.Object(
        {
            a: CS.String(true),
            b: CS.Number(true),
        },
        true,
    );
    type objType = typeof obj.type;

    checkType<objType>({ a: 'abcd', b: 2 });
    expect(obj.getJsonSchema()).toEqual({
        type: 'object',
        additionalProperties: false,
        properties: {
            a: {
                type: 'string',
                transform: ['trim'],
            },
            b: {
                type: 'number',
            },
        },
        required: ['a', 'b'],
    });
});

it('works with object schema and optional', () => {
    const obj = CS.Object(
        {
            a: CS.String(true),
            b: CS.String(false),
        },
        true,
    );
    type objType = typeof obj.type;

    checkType<objType>({
        a: 'b',
        b: undefined,
    });
    checkType<objType>({
        a: 'b',
        b: 'b',
    });

    expect(obj.getJsonSchema()).toEqual({
        type: 'object',
        additionalProperties: false,
        properties: {
            a: {
                type: 'string',
                transform: ['trim'],
            },
            b: {
                type: 'string',
                transform: ['trim'],
            },
        },
        required: ['a'],
    });
});

it('works with object schema with all optional', () => {
    const obj = CS.Object(
        {
            a: CS.String(false),
            b: CS.String(false),
        },
        true,
    );
    type objType = typeof obj.type;

    checkType<objType>({
        a: undefined,
        b: undefined,
    });
    checkType<objType>({
        a: 'b',
        b: 'b',
    });

    expect(obj.getJsonSchema()).toEqual({
        type: 'object',
        additionalProperties: false,
        properties: {
            a: {
                type: 'string',
                transform: ['trim'],
            },
            b: {
                type: 'string',
                transform: ['trim'],
            },
        },
    });
});

it('works with mergeobjects schema', () => {
    const obj = CS.MergeObjects(
        CS.Object(
            {
                a: CS.String(true),
            },
            true,
        ),
        CS.Object(
            {
                b: CS.String(false),
            },
            true,
        ),
        true,
    );
    type objType = typeof obj.type;

    checkType<objType>({
        a: 'a',
        b: undefined,
    });
    checkType<objType>({
        a: 'b',
        b: 'b',
    });

    expect(obj.getJsonSchema()).toEqual({
        type: 'object',
        additionalProperties: false,
        properties: {
            a: {
                type: 'string',
                transform: ['trim'],
            },
            b: {
                type: 'string',
                transform: ['trim'],
            },
        },
        required: ['a'],
    });
});

it('works with mergeobjects schema and no required fields', () => {
    const obj = CS.MergeObjects(
        CS.Object(
            {
                a: CS.String(false),
            },
            true,
        ),
        CS.Object(
            {
                b: CS.String(false),
            },
            true,
        ),
        true,
    );
    type objType = typeof obj.type;

    checkType<objType>({
        a: undefined,
        b: undefined,
    });
    checkType<objType>({
        a: 'b',
        b: 'b',
    });

    expect(obj.getJsonSchema()).toEqual({
        type: 'object',
        additionalProperties: false,
        properties: {
            a: {
                type: 'string',
                transform: ['trim'],
            },
            b: {
                type: 'string',
                transform: ['trim'],
            },
        },
        required: [],
    });
});

it('works with array schema', () => {
    const arr = CS.Array(CS.String(true), true);
    type arrType = typeof arr.type;

    checkType<arrType>(['abcd', 'efgh']);
    expect(arr.getJsonSchema()).toEqual({
        type: 'array',
        items: {
            type: 'string',
            transform: ['trim'],
        },
    });
});

it('works with enum schema', () => {
    const cs = CS.Enum(['a' as 'a', 'b' as 'b'], true);
    type csType = typeof cs.type;

    checkType<csType>('a');
    checkType<csType>('b');

    expect(cs.getJsonSchema()).toEqual({
        type: 'string',
        enum: ['a', 'b'],
    });
});
