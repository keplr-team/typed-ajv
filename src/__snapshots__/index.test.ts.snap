// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typed-ajv Any() works with any schema 1`] = `Object {}`;

exports[`typed-ajv AnyOf() generates an anyOf schema 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "foo": Object {
      "anyOf": Array [
        Object {
          "transform": Array [
            "trim",
          ],
          "type": "string",
        },
        Object {
          "type": "boolean",
        },
      ],
    },
  },
  "required": Array [
    "foo",
  ],
  "type": "object",
}
`;

exports[`typed-ajv Array() works with array schema 1`] = `
Object {
  "items": Object {
    "transform": Array [
      "trim",
    ],
    "type": "string",
  },
  "type": "array",
}
`;

exports[`typed-ajv Array() works with array schema and opts 1`] = `
Object {
  "items": Object {
    "transform": Array [
      "trim",
    ],
    "type": "string",
  },
  "minItems": 1,
  "type": "array",
  "uniqueItems": true,
}
`;

exports[`typed-ajv Boolean() works with boolean schema 1`] = `
Object {
  "type": "boolean",
}
`;

exports[`typed-ajv Const() works with const schema 1`] = `
Object {
  "const": 42,
}
`;

exports[`typed-ajv Enum() works with enum schema 1`] = `
Object {
  "enum": Array [
    "a",
    "b",
  ],
  "type": "string",
}
`;

exports[`typed-ajv Integer() works with integer schema 1`] = `
Object {
  "type": "integer",
}
`;

exports[`typed-ajv Integer() works with integer schema and options 1`] = `
Object {
  "maximum": 4,
  "minimum": 1,
  "type": "integer",
}
`;

exports[`typed-ajv MergeObjects works with mergeobjects schema 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "a": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
    "b": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
  },
  "required": Array [
    "a",
  ],
  "type": "object",
}
`;

exports[`typed-ajv MergeObjects works with mergeobjects schema and no required fields 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "a": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
    "b": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
  },
  "required": Array [],
  "type": "object",
}
`;

exports[`typed-ajv Null() works with null schema 1`] = `
Object {
  "type": "null",
}
`;

exports[`typed-ajv Number() works with number schema 1`] = `
Object {
  "type": "number",
}
`;

exports[`typed-ajv Number() works with number schema and options 1`] = `
Object {
  "exclusiveMaximum": 5,
  "exclusiveMinimum": 3,
  "type": "number",
}
`;

exports[`typed-ajv Object() works with object schema 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "a": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
    "b": Object {
      "type": "number",
    },
  },
  "required": Array [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`typed-ajv Object() works with object schema and optional 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "a": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
    "b": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
  },
  "required": Array [
    "a",
  ],
  "type": "object",
}
`;

exports[`typed-ajv Object() works with object schema and options 1`] = `
Object {
  "additionalProperties": true,
  "description": "object",
  "properties": Object {
    "a": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
    "b": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
  },
  "required": Array [
    "a",
  ],
  "type": "object",
}
`;

exports[`typed-ajv Object() works with object schema with all optional 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "a": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
    "b": Object {
      "transform": Array [
        "trim",
      ],
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`typed-ajv String() works with string schema 1`] = `
Object {
  "transform": Array [
    "trim",
  ],
  "type": "string",
}
`;

exports[`typed-ajv String() works with string schema and options 1`] = `
Object {
  "maxLength": 10,
  "minLength": 1,
  "transform": Array [
    "trim",
  ],
  "type": "string",
}
`;

exports[`typed-ajv Unknown() works with unknown schema 1`] = `Object {}`;
